### PR TITLE
[Woo POS] Create onboarding adaptor

### PR DESCRIPTION
**Pull request by joshheald**
Part of: #12866

## Description

This adds a reimplementation of the onboarding presenter, adapting the output to be more suitable for use with SwiftUI.

## Testing instructions

This is not in use yet, so no testing other than checking that CI passes

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsPresenting.swift [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import UIKit
+
+protocol CardPresentPaymentsPresenting {
```
Comment (jaclync) [inline WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsPresenting.swift]: ❓  This doesn't seem to be used in this PR so far, and looks identical to `ViewControllerPresenting` in https://github.com/woocommerce/woocommerce-ios/pull/12896. How do you plan to use this protocol?

**Review by jaclync — COMMENTED** [review]
Not sure if it's ready for a look, just leaving a passing question!

File: WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsPresenting.swift (hunk 1 above)
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/CardPresentPayments/CardPresentPaymentsPresenting.swift]: Oops. I just left it in having planned to use the generic one. Thanks!

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    private let onboardingViewModel: CardPresentPaymentsOnboardingViewModel
+
+    private var readinessSubscription: AnyCancellable?
+
+    private(set) var onboardingScreenViewModelPublisher: AnyPublisher<CardPresentPaymentOnboardingPresentationEvent, Never>
+
+    private let onboardingScreenViewModelSubject: PassthroughSubject<CardPresentPaymentOnboardingPresentationEvent, Never> = PassthroughSubject()
+
+    init(stores: StoresManager = ServiceLocator.stores) {
+        self.stores = stores
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift]: nit: does `stores` need to be held by the adaptor? from the implementation it looks like it's just passed to `CardPresentPaymentsOnboardingUseCase` and `CardPresentPaymentsReadinessUseCase`, but maybe there are other plans to use it in a later task?

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    private let onboardingUseCase: CardPresentPaymentsOnboardingUseCase
+
+    private let readinessUseCase: CardPresentPaymentsReadinessUseCase
+
+    private let onboardingViewModel: CardPresentPaymentsOnboardingViewModel
+
+    private var readinessSubscription: AnyCancellable?
+
+    private(set) var onboardingScreenViewModelPublisher: AnyPublisher<CardPresentPaymentOnboardingPresentationEvent, Never>
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift]: super nit: if we don't plan to change the publisher
```suggestion
    let onboardingScreenViewModelPublisher: AnyPublisher<CardPresentPaymentOnboardingPresentationEvent, Never>
```

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+        readinessSubscription = readinessUseCase.$readiness
+            .subscribe(on: DispatchQueue.main)
+            .sink(receiveValue: { [weak self] readiness in
+                guard let self,
+                      case .ready = readiness else {
+                    return
+                }
+
+                onboardingScreenViewModelSubject.send(.onboardingComplete)
```
Comment (jaclync) [inline WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift:58]: I checked the UIKit version, so nice with the publisher!

**Review by jaclync — APPROVED** [review]
LGTM! :shipit:

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift (hunk 4 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift:58]: I played with it more in the experiment and I think we may have some issues actually using it to display the UI, so we may still need to rely on the POS feature gate requiring onboarding completion for a while.

It was really strange – the underlying viewmodel's state was updating, and the onboarding view's `body` was called and went through the switch case to choose the right screen, but it never changed what was displayed.

That's a problem for another day though!

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift (hunk 2 above)
Comment (joshheald) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift]: No, you're right, I'll remove the property and we can add it if we need it in future

File: WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift (hunk 4 above)
Comment (jaclync) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift:58]: Could you share the steps to repro this if it's reproducible? We can log this first and look into this after the happy path.

Comment (joshheald) [reply WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentOnboardingAdaptor.swift:58]: Quite a few ways, but one repro was along the lines of

1. Disable WooPayments on your store
2. Edit the POS eligibility checks to make sure you still see the button
3. Launch the app
4. Go to `Menu > POS Mode`
5. Try to connect the reader
6. Observe that the "WooPayments is disabled" onboarding step is correctly shown
7. Tap "Activate WooPayments"
8. Observe that it never changes what's shown on screen.

Breakpoints can show it going through the OnboardingView switch statement

**Pull request merged**